File: src/main/java/com/machfour/macros/cli/MealPrinter.java
package com.machfour.macros.cli;

import com.machfour.macros.core.Column;
import com.machfour.macros.objects.FoodPortion;
import com.machfour.macros.objects.Meal;
import com.machfour.macros.objects.NutritionData;
import com.machfour.macros.objects.QtyUnit;
import com.machfour.macros.util.PrintFormatting;
import com.machfour.macros.util.StringJoiner;
import com.machfour.macros.util.UnicodeUtils;
import org.jetbrains.annotations.NotNull;

import java.io.PrintStream;
import java.util.*;

import static com.machfour.macros.core.Schema.NutritionDataTable.*;

class MealPrinter {
    private static final String columnSep = "   ";
    private static final List<Column<NutritionData, Double>> conciseTableCols;
    private static final List<Column<NutritionData, Double>> verboseTableCols;

    static {
        conciseTableCols = Arrays.asList(
                CALORIES
                , PROTEIN
                , FAT
                , CARBOHYDRATE
        );
        verboseTableCols = Arrays.asList(
                CALORIES
                , PROTEIN
                , FAT
                , SATURATED_FAT
                , CARBOHYDRATE
                , SUGAR
                , FIBRE
                , SODIUM
                , CALCIUM
        );
    }

    private static int sum(Collection<Integer> sums) {
        int sum = 0;
        for (Integer i : sums) {
            sum += i;
        }
        return sum;
    }

    /*
     * Prints a row of a table (helper method for printMeal())
     */
    private static void printRow(List<String> row, List<Integer> widths, List<Boolean> rightAlign, String sep, PrintStream out) {
        assert (row.size() == widths.size() && row.size() == rightAlign.size());
        for (int i = 0; i < row.size(); ++i) {
            String align = rightAlign.get(i) ? "" : "-";
            int width = widths.get(i);
            String text = row.get(i);
            // fullwidth characters are printed as two characters in a terminal, so we should reduce width by the
            // number of fullwidth characters in the string
            int numDoubleWidthChars = UnicodeUtils.countDoubleWidthChars(text);

            // displayed length of text appears to be length() + numDoubleWidthChars characters long.
            // Equivalently, we can reduce width by this amount, to get the printing right
            width = Math.max(width - numDoubleWidthChars, 0);


            // prevent long strings from overrunning the width:
            // replace "This is a really long string"
            // with    "This is a really lo.."
            if (text.length() > width) {
                int newWidth = Math.max(width - 2, 0);
                // TODO with double width chars, this may reduce by too much
                text = text.substring(0, newWidth - 2) + "..";
            }
            //String widthStr = String.valueOf(width);
            out.printf("%" + align + width + "s%s", text, sep);
        }
        out.println();
    }

    private static List<String> nutritionDataToRow(String name, NutritionData nd, double qty, QtyUnit unit, boolean verbose)  {
        List<Column<NutritionData, Double>> nutrientColumns = verbose ? verboseTableCols : conciseTableCols;
        List<String> row = new ArrayList<>(nutrientColumns.size() + 2);
        // add food name
        row.add(name);
        // add nutrients, formatting to be the appropriate width
        for (Column<NutritionData, Double> nutrient : nutrientColumns) {
            Double value = nd.amountOf(nutrient);
            row.add(PrintFormatting.formatQuantity(value, verbose));
        }
        // add quantity and unit
        row.add(PrintFormatting.formatQuantity(qty, unit, PrintFormatting.servingWidth));
        return row;
    }

    static void printMeal(@NotNull Meal meal, boolean verbose, @NotNull PrintStream out) {
        List<Column<NutritionData, Double>> nutrientCols = verbose ? verboseTableCols : conciseTableCols;
        //Columns: first the food name, then one for each nutrient, then quantity/serving
        int numCols = 1 + nutrientCols.size() + 1;
        // holds the meal name and labels for each nutrient column
        List<String> headingRow = new ArrayList<>(numCols);
        List<Integer> rowWidths = new ArrayList<>(numCols);
        List<Boolean> rightAlign = new ArrayList<>(numCols);
        // first column has meal name (heading) and food names for other rows
        headingRow.add(meal.getName());
        rowWidths.add(PrintFormatting.nameWidth);
        rightAlign.add(false);
        // next columns have names for each nutrient (heading) then corresponding data
        for (Column<NutritionData, Double> col : nutrientCols) {
            if (verbose) {
                headingRow.add(PrintFormatting.longerNames.get(col));
                rowWidths.add(PrintFormatting.longDataWidth);
            } else {
                headingRow.add(PrintFormatting.briefNames.get(col));
                rowWidths.add(PrintFormatting.shortDataWidth);
            }
            rightAlign.add(true);
        }
        // last column is quantity, so is a bit longer
        headingRow.add(PrintFormatting.briefNames.get(QUANTITY));
        rowWidths.add(PrintFormatting.servingWidth);
        rightAlign.add(true);

        // row separator spans all columns plus each separator, but we discount the space
        // after the last separator
        int rowSepLength = sum(rowWidths) + rowWidths.size()*columnSep.length() - 1;
        String rowSeparator = StringJoiner.of(Collections.nCopies(rowSepLength, "=")).join();

        printRow(headingRow, rowWidths, rightAlign, columnSep, out);
        out.println(rowSeparator);
        // now we get to the actual data
        List<List<String>> dataRows = new ArrayList<>();
        for (FoodPortion fp : meal.getFoodPortions()) {
            String name = fp.getFood().getMediumName();
            NutritionData nd = fp.getNutritionData();
            dataRows.add(nutritionDataToRow(name, nd, fp.getQuantity(), fp.getQtyUnit(), verbose));
        }
        for (List<String> row : dataRows) {
            printRow(row, rowWidths, rightAlign, columnSep, out);
        }
        // now print total
        out.println(rowSeparator);
        String totalName = String.format("Total for %s", meal.getName());
        NutritionData totalNd = meal.getNutritionTotal();
        // for total data, just use the quantity and unit from the sum
        List<String> totalRow = nutritionDataToRow(totalName, totalNd, totalNd.getQuantity(), totalNd.qtyUnit(), verbose);
        printRow(totalRow, rowWidths, rightAlign, columnSep, out);
    }

    static void printMeals(Collection<Meal> meals, PrintStream out, boolean verbose, boolean per100, boolean grandTotal) {
        out.println("============");
        out.println("Meal totals:");
        out.println("============");
        out.println();
        for (Meal m : meals) {
            if (m.getFoodPortions().isEmpty()) {
                out.println("Meal " + m.getName() + " has no recorded data");
            } else {
                printMeal(m, verbose, out);
                out.println();
                if (per100) {
                    out.println("== Nutrient total per 100g ==");
                    CliUtils.printPer100g(m.getNutritionTotal(), verbose, out);
                    out.println("=============================");
                    out.println();
                }
            }
        }
        if (grandTotal) {
            List<NutritionData> allNutData = new ArrayList<>();
            for (Meal m : meals) {
                allNutData.add(m.getNutritionTotal());
            }
            NutritionData totalNutData = NutritionData.sum(allNutData);
            out.println("====================");
            out.println("Total for all meals:");
            out.println("====================");
            CliUtils.printNutritionData(totalNutData, verbose, out);
            out.println();
            CliUtils.printEnergyProportions(totalNutData, verbose, out);
        }
    }

    static void printMeals(Collection<Meal> meals, PrintStream out) {
        printMeals(meals, out, false, false, true);
    }
}

File: src/main/kotlin/com/machfour/macros/cli/utils/MealPrinter.kt
package com.machfour.macros.cli.utils

import com.machfour.macros.names.EnglishColumnNames
import com.machfour.macros.objects.*
import com.machfour.macros.objects.Unit
import com.machfour.macros.objects.inbuilt.Nutrients
import com.machfour.macros.util.PrintFormatting
import com.machfour.macros.util.PrintFormatting.formatQuantity
import com.machfour.macros.util.StringJoiner
import com.machfour.macros.util.UnicodeUtils
import com.machfour.macros.util.UnicodeUtils.countDoubleWidthChars
import java.io.PrintStream
import kotlin.collections.ArrayList

object MealPrinter {
    private const val columnSep = "   "
    private val conciseTableCols = listOf(
          Nutrients.ENERGY
        , Nutrients.PROTEIN
        , Nutrients.FAT
        , Nutrients.CARBOHYDRATE
    )

    private val verboseTableCols = listOf(
          Nutrients.ENERGY
        , Nutrients.PROTEIN
        , Nutrients.FAT
        , Nutrients.SATURATED_FAT
        , Nutrients.CARBOHYDRATE
        , Nutrients.SUGAR
        , Nutrients.FIBRE
        , Nutrients.SODIUM
        , Nutrients.CALCIUM
    )

    /*
     * Prints a row of a table (helper method for printMeal())
     */
    private fun printRow(row: List<String>, widths: List<Int>, rightAlign: List<Boolean>, sep: String, out: PrintStream) {
        assert(row.size == widths.size && row.size == rightAlign.size)
        for (i in row.indices) {
            out.print(UnicodeUtils.formatString(row[i], widths[i], !rightAlign[i]) + sep)
        }
        out.println()
    }

    private fun nutritionDataToRow(name: String, nd: NutritionData, qty: Double, unit: Unit, verbose: Boolean): List<String> {
        val nutrientColumns = if (verbose) verboseTableCols else conciseTableCols
        val row: MutableList<String> = ArrayList(nutrientColumns.size + 2)
        // add food name
        row += name
        val nutrientWidth = if (verbose) PrintFormatting.longDataWidth else PrintFormatting.shortDataWidth
        // add nutrients, formatting to be the appropriate width
        for (nutrient in nutrientColumns) {
            row += formatQuantity(nd, nutrient, width = nutrientWidth, unitWidth = 2, withDp = verbose)
        }
        // add quantity and unit
        row.add(formatQuantity(qty, unit, width = PrintFormatting.servingWidth, unitWidth = 2))
        return row
    }

    fun printMeal(meal: Meal, verbose: Boolean, out: PrintStream) {
        val nutrientCols = if (verbose) verboseTableCols else conciseTableCols
        //Columns: first the food name, then one for each nutrient, then quantity/serving
        val numCols = 1 + nutrientCols.size + 1
        // holds the meal name and labels for each nutrient column
        val headingRow: MutableList<String> = ArrayList(numCols)
        val rowWidths: MutableList<Int> = ArrayList(numCols)
        val rightAlign: MutableList<Boolean> = ArrayList(numCols)
        // first column has meal name (heading) and food names for other rows
        headingRow.add(meal.name)
        rowWidths.add(PrintFormatting.nameWidth)
        rightAlign.add(false)
        // next columns have names for each nutrient (heading) then corresponding data
        for (col in nutrientCols) {
            if (verbose) {
                headingRow.add(EnglishColumnNames.longerNutrientNames.getValue(col))
                rowWidths.add(PrintFormatting.longDataWidth)
            } else {
                headingRow.add(EnglishColumnNames.briefNutrientNames.getValue(col))
                rowWidths.add(PrintFormatting.shortDataWidth)
            }
            rightAlign.add(true)
        }
        // last column is quantity, so is a bit longer
        headingRow.add(EnglishColumnNames.briefNutrientNames.getValue(Nutrients.QUANTITY))
        rowWidths.add(PrintFormatting.servingWidth)
        rightAlign.add(true)

        // row separator spans all columns plus each separator, but we discount the space
        // after the last separator
        val rowSepLength = rowWidths.sum() + rowWidths.size * columnSep.length - 1
        val rowSeparator = StringJoiner.of("=").copies(rowSepLength).join()
        printRow(headingRow, rowWidths, rightAlign, columnSep, out)
        out.println(rowSeparator)
        // now we get to the actual data
        val dataRows: MutableList<List<String>> = ArrayList()
        for (fp in meal.getFoodPortions()) {
            val name = fp.food.mediumName
            val nd = fp.nutritionData.withDefaultUnits()
            dataRows.add(nutritionDataToRow(name, nd, fp.quantity, fp.qtyUnit, verbose))
        }
        for (row in dataRows) {
            printRow(row, rowWidths, rightAlign, columnSep, out)
        }
        // now print total
        out.println(rowSeparator)
        val totalName = "Total for ${meal.name}"
        val totalNd = meal.nutritionTotal
        // for total data, just use the quantity and unit from the sum
        val totalRow = nutritionDataToRow(totalName, totalNd, totalNd.quantity, totalNd.qtyUnit, verbose)
        printRow(totalRow, rowWidths, rightAlign, columnSep, out)
    }

    fun printMeals(meals: Collection<Meal>, out: PrintStream, verbose: Boolean = false, per100: Boolean = false, grandTotal: Boolean = true) {
        out.println("============")
        out.println("Meal totals:")
        out.println("============")
        out.println()
        for (m in meals) {
            if (m.getFoodPortions().isEmpty()) {
                out.println("Meal ${m.name} has no recorded data")
            } else {
                printMeal(m, verbose, out)
                out.println()
                if (per100) {
                    out.println("== Nutrient total per 100g ==")
                    CliUtils.printPer100g(m.nutritionTotal, verbose, out)
                    out.println("=============================")
                    out.println()
                }
            }
        }
        if (grandTotal) {
            val allNutData: MutableList<NutritionData> = ArrayList()
            for (m in meals) {
                allNutData.add(m.nutritionTotal)
            }
            val totalNutData = NutritionCalculations.sum(allNutData)
            out.println("====================")
            out.println("Total for all meals:")
            out.println("====================")
            CliUtils.printNutritionData(totalNutData, verbose, out)
            out.println()
            CliUtils.printEnergyProportions(totalNutData, verbose, out)
        }
    }

}